### PR TITLE
Generalize ML dataset parsing

### PR DESCRIPTION
**Pull request by kaminow**
The original goal was to make the regexes that are used to extract crystal structure name and compound id from filenames not hardcoded. Other stuff that was changed along the way:
- No longer automatically filter out experimental compounds that don't have structural data when creating a GraphDataset (now an option in the `build_dataset` function)
- No longer automatically remove compounds with NaN values for "pIC50_range" or "pIC50_stderr" (now options in the `load_exp_data` function)

### CODE REVIEW

File: asapdiscovery-data/asapdiscovery/data/utils.py [hunk 1]
```diff
@@ -16,6 +16,11 @@
     ExperimentalCompoundDataUpdate,
 )
 
+# Not sure if this is the right place for these
+# Regex patterns for extracting Mpro dataset ID and Moonshot CDD style compound ID
+MOONSHOT_CDD_ID_REGEX = r"[A-Z]{3}-[A-Z]{3}-[0-9a-z]+-[0-9]+"
+MPRO_ID_REGEX = r"Mpro-.*?_[0-9][A-Z]"
```
[the remark this answers is not in the document]
Comment (JenkeScheen) [reply asapdiscovery-data/asapdiscovery/data/utils.py:22]: FWIW, we should be changing our mindset away from 'mpro' as we're about to have other targets pour in which is going to mess up all these variable names.

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 2]
```diff
@@ -28,6 +33,12 @@ def build_dataset(
         Which model to create. Current options are ["2d", "schnet", "e3nn"]
     exp_fn : str
         JSON file giving experimental results
+    xtal_pat : str, default=r"[^/]+$"
+        Regex pattern for extracting crystal structure ID from filename. Defaults to
+        taking the entire file basename (on *nix platforms)
+    compound_pat : str, default=r"[^/]+$"
```
Comment (JenkeScheen) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:39]: for these kind of regex strings in docstrings, would you be able to give a brief explanation of what they do? I'm very bad at reading them. ChatGPT thinks this one's:
> matches any string of characters, where all letters and numbers are at least one character long.

File: asapdiscovery-data/asapdiscovery/data/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-data/asapdiscovery/data/utils.py:22]: maybe we could have a YAML file or something storing different regexes for different projects?

File: asapdiscovery-ml/asapdiscovery/ml/utils.py (hunk 2 above)
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/utils.py:40]: I updated the description, hopefully it's a bit more clear now

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 3]
```diff
@@ -56,45 +77,57 @@ def build_dataset(
...
+    exp_data, exp_compounds = load_exp_data(
+        exp_fn,
+        achiral=achiral,
+        return_compounds=True,
+        check_range_nan=check_range_nan,
+        check_stderr_nan=check_stderr_nan,
+    )
+
+    # Parse structure filenames
+    if (model_type.lower() != "2d") or str_only:
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:91]: I have been moving some instances of `2d` to be `GAT` does this change make sense/ are you happy with it? If so we could make that change or I can revert some of my changes.

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 4]
```diff
@@ -10,12 +10,17 @@ def build_dataset(
     in_files,
     model_type,
     exp_fn,
+    xtal_pat=r"[^/]+$",
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:13]: Same here I would do the filtering outside the function by whatever means nessecary and then accept iterables of the relevant data, that way regex is not absolutely required.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few comments @kaminow, let me know what you think and we can discuss.:)

File: asapdiscovery-ml/asapdiscovery/ml/utils.py (hunk 3 above)
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/utils.py]: yeah I think this makes sense. I've switched everything (I think) over in [e37f0f2](https://github.com/choderalab/asapdiscovery/pull/198/commits/e37f0f2d5607ac2a11bf63f6ec0c01a9c10abd55)

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 5]
```diff
@@ -356,17 +367,35 @@ def init(args, rank=False):
...
+        # Using the stderr information in loss calculations, so need to include both
+        check_range_nan = True
+        check_stderr_nan = True
+
     # Load full dataset
     ds, exp_data = build_dataset(
         in_files=args.i,
         model_type=args.model,
         exp_fn=args.exp,
+        xtal_pat=args.xtal_regex if args.xtal_regex else MPRO_ID_REGEX,
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:389]: this also definitely makes sense, I'll rework the `build_dataset` function a bit

**Comment by kaminow** [comment]
@hmacdope I've made some changes based on your suggestions:
- `build_dataset` now takes lists of filenames and compounds (now optional as well, since we don't actually need them for the `GraphDataset`)
- moved the regex parsing functionality to its own function in `asapdiscovery.data.utils`
- `train.py` now uses this function, as well as some other code that used to be in the `build_dataset` function
thanks again for the review, lmk if you think there's anything else that you think should be changed/added!

File: asapdiscovery-data/asapdiscovery/data/utils.py [hunk 6]
```diff
@@ -159,6 +164,64 @@ def edit_pdb_file(
     print(f"Wrote {pdb_out}", flush=True)
 
 
+def extract_compounds_from_filenames(fn_list, xtal_pat, compound_pat, fail_val=None):
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/utils.py:167]: Much prefer it this way, well done!

File: asapdiscovery-data/asapdiscovery/data/utils.py [hunk 7]
```diff
@@ -159,6 +164,64 @@ def edit_pdb_file(
...
+        List of (xtal, compound_id)
+    """
+    import re
+
+    if callable(xtal_pat):
+        # Just use the passed function
+        xtal_func = xtal_pat
+    else:
+        # Construct a function for re searching
+        def xtal_func(fn):
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/utils.py:199]: Could seperate this into a function itself, `construct_regex_function` but up to you.

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 8]
```diff
@@ -356,17 +371,61 @@ def init(args, rank=False):
...
+    if args.i:
+        # Parse compounds from args.i
+        if os.path.isdir(args.i):
+            all_fns = glob(f"{args.i}/*complex.pdb")
+        else:
+            all_fns = glob(args.i)
+        check_filelist_has_elements(all_fns, "ml_dataset")
+
+        # Parse compound filenames
+        xtal_regex = args.xtal_regex if args.xtal_regex else MPRO_ID_REGEX
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:397]: This is much better IMO :)

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 9]
```diff
@@ -7,27 +7,33 @@
...
     exp_fn,
+    all_fns=[],
+    compounds=[],
     achiral=False,
     cache_fn=None,
     grouped=False,
     lig_name="LIG",
     num_workers=1,
     rank=False,
+    str_only=False,
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:20]: str_only is a bit confusing as I think `string`, perhaps `structure_only`?

**Review by hmacdope — APPROVED** [review]
Looks good @kaminow!  :)

**Comment by hmacdope** [comment]
I have re-triggered CI, (looked like `actions` downtime) lets see what happens

File: asapdiscovery-ml/asapdiscovery/ml/utils.py (hunk 9 above)
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/utils.py]: that's reasonable. done in [ad31008](https://github.com/choderalab/asapdiscovery/pull/198/commits/ad31008b71b96fbf8e4fd1c56093840d822800cf)

File: asapdiscovery-data/asapdiscovery/data/utils.py (hunk 7 above)
Comment (kaminow) [reply asapdiscovery-data/asapdiscovery/data/utils.py]: done in [1366284](https://github.com/choderalab/asapdiscovery/pull/198/commits/13662848623d9d9057b1d06bcfd862bbec37cd1a)

**Comment by kaminow** [comment]
@hmacdope can you give these new commits one last quick look over? feel to to merge if everything looks good and CI passes, or I can take care of it in the morning. thanks!

**Comment by hmacdope** [comment]
@kaminow yep for sure. :)

**Review by hmacdope — APPROVED** [review]
LGTM

**Pull request merged**